File: 求从数组中取出n个元素的所有组合/combine.js

var combine_recursive = function(arr,result,start,single_result, index,n ,arr_len){
    for(let ct = start;ct < arr_len-index+1;ct++){
        single_result[n-index] = ct;
        if(index===1){
            result.push(single_result.map(item=>arr[item]));
        }
        else{
            combine_recursive(arr,result,ct+1,single_result,index-1,n,arr_len);
        }
    }
};

var do_combine = function (arr) {
    let result = [];
    for(let n = 1;n<=arr.length;n++){
        let start = 0;
        let single_result = [];
        combine_recursive(arr,result,start,single_result,n,n,arr.length);
        // result.push(single_result);
    }
    return result;
};


//测试代码
var combines_multi;
tryList.forEach(t=>{
    if(t.quizTypeId === "itt004" && typeof(t.rightAnswer) === 'undefined'){
        combines_multi = do_combine(t.options);
    }
});

File: 求从数组中取出n个元素的所有组合/combine.c
#include<stdio.h>
void combine (int *arr,int start,int *result,int index,int n,int arr_len)
{
    for(int ct = start;ct < arr_len-index+1;ct++){
        result[n-index] = ct;
        if(index==1){
            for(int j = 0;j<n;j++) {
                printf("%d\t", arr[result[j]]);
            }
            printf("\n");
        }
        else{
            combine(arr,ct+1,result,index-1,n,arr_len);
        }
    }
}

int main()
{
	int Num = 6;
    int arr[Num]={1,2,3,4,5,6};
    int result[Num]={};
    int arr_len = sizeof(arr)/sizeof(int);

    for(int n = 1;n<=arr_len;n++){
        int start = 0;
        int index = n;
        combine(arr,start,result,index,n,arr_len);
    }

}